### PR TITLE
AST: Cache SuperclassDeclRequest

### PR DESCRIPTION
**Pull request by slavapestov**
This fixes a significant build time regression since 4.2. Fixes rdar://problem/47305605.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -3562,11 +3562,16 @@ class ClassDecl final : public NominalTypeDecl {
...
 
   struct {
+    /// The superclass decl and a bit to indicate whether the
+    /// superclass was computed yet or not.
+    llvm::PointerIntPair<ClassDecl *, 1, bool> SuperclassDecl;
+
     /// The superclass type and a bit to indicate whether the
     /// superclass was computed yet or not.
-    llvm::PointerIntPair<Type, 1, bool> Superclass;
+    llvm::PointerIntPair<Type, 1, bool> SuperclassType;
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: Memory-usage-wise, do we need both? Seems like storing `PointerUnion<ClassDecl *, Type>` inside the PointerIntPair would be sufficient.

Comment (DougGregor) [reply include/swift/AST/Decl.h]: The cached superclass `Decl` and superclass `Type ` are handled by different conceptual layers (name lookup vs. type resolution), so they're really different pieces of state. I don't think the benefits of folding them together are worth the risk of unpredictable behavior.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (slavapestov) [reply include/swift/AST/Decl.h]: Yeah. Even in a codebase with 10,000 classes and protocols, you're looking at an 80kb savings.

**Review by DougGregor — APPROVED** [review]
LGTM, thank you! I'd like to poke at using the evaluator's caching, but that's more plumbing that we don't need to do at this moment.

**Pull request merged**